tests/test_dtypes.py

```diff
@@ -0,0 +1,202 @@
+import numpy as np
+import pytest
+import sympy
+
+from devito import Constant, Eq, Function, Grid, Operator
+from devito.ir.iet.nodes import Dereference
+from devito.passes.iet.langbase import LangBB
+from devito.passes.iet.languages.C import CBB
+from devito.passes.iet.languages.openacc import AccBB
+from devito.passes.iet.languages.openmp import OmpBB
+from devito.symbolics.extended_dtypes import Float16P
+from devito.types.basic import Basic, Scalar, Symbol
+from devito.types.dimension import Dimension, Spacing
+
+# Mappers for language-specific types and headers
+_languages: dict[str, type[LangBB]] = {
+    'C': CBB,
+    'openmp': OmpBB,
+    'openacc': AccBB
+}
+
+
+def _get_language(language: str, **_) -> type[LangBB]:
+    """
+    Gets the language building block type from parametrized kwargs.
+    """
+
+    return _languages[language]
+
+
+def _config_kwargs(platform: str, language: str, compiler: str) -> dict[str, str]:
+    """
+    Generates kwargs for Operator to test language-specific behavior.
+    """
+
+    return {
+        'platform': platform,
+        'language': language,
+        'compiler': compiler
+    }
+
+
+# List of pararmetrized operator kwargs for testing language-specific behavior
+_configs: list[dict[str, str]] = [
+    _config_kwargs(*cfg) for cfg in [
+        ('cpu64', 'C', 'gcc'),
+        ('cpu64', 'openmp', 'gcc'),
+        ('nvidiaX', 'openmp', 'nvc'),
+        ('nvidiaX', 'openacc', 'nvc')
+    ]
+]
+
+
+@pytest.mark.parametrize('dtype', [np.float16, np.complex64, np.complex128])
+@pytest.mark.parametrize('kwargs', _configs)
+def test_dtype_mapping(dtype: np.dtype, kwargs: dict[str, str]) -> None:
+    """
+    Tests that half and complex floats' dtypes result in the correct type
+    strings in generated code.
+    """
+
+    # Retrieve the language-specific type mapping
+    lang_types: dict[np.dtype, type] = _get_language(**kwargs).get('types')
+
+    # Set up an operator
+    grid = Grid(shape=(3, 3))
+    x: Dimension
+    y: Dimension
```

Comment:
These can be dropped later when ready to merge, I guess they are helpful for you now